x/auth/tx/builder.go

```diff
@@ -61,14 +70,84 @@ func newBuilder(cdc codec.Codec) *wrapper {
 			},
 		},
 	}
+	return w
 }
 
 func (w *wrapper) GetMsgs() []sdk.Msg {
 	return w.tx.GetMsgs()
 }
 
+func (w *wrapper) GetMsgsV2() ([]protov2.Message, error) {
+	if w.msgsV2 == nil {
+		w.initSignersAndMsgsV2()
+	}
+
+	return w.msgsV2, w.err
+}
+
 func (w *wrapper) ValidateBasic() error {
-	return w.tx.ValidateBasic()
+	if w.tx == nil {
+		return fmt.Errorf("bad Tx")
+	}
+
+	body := w.tx.Body
+	if body == nil {
+		return fmt.Errorf("missing TxBody")
+	}
+
+	authInfo := w.tx.AuthInfo
+	if authInfo == nil {
+		return fmt.Errorf("missing AuthInfo")
+	}
+
+	fee := authInfo.Fee
+	if fee == nil {
+		return fmt.Errorf("missing fee")
+	}
+
+	if fee.GasLimit > tx.MaxGasWanted {
+		return errorsmod.Wrapf(
+			sdkerrors.ErrInvalidRequest,
+			"invalid gas supplied; %d > %d", fee.GasLimit, tx.MaxGasWanted,
+		)
+	}
+
+	if fee.Amount.IsAnyNil() {
+		return errorsmod.Wrapf(
+			sdkerrors.ErrInsufficientFee,
+			"invalid fee provided: null",
+		)
+	}
+
+	if fee.Amount.IsAnyNegative() {
+		return errorsmod.Wrapf(
+			sdkerrors.ErrInsufficientFee,
+			"invalid fee provided: %s", fee.Amount,
+		)
+	}
+
+	if fee.Payer != "" {
+		_, err := sdk.AccAddressFromBech32(fee.Payer)
+		if err != nil {
+			return errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "Invalid fee payer address (%s)", err)
+		}
+	}
+
+	sigs := w.tx.Signatures
+
+	if len(sigs) == 0 {
+		return sdkerrors.ErrNoSignatures
+	}
+
+	signers, _ := w.GetSigners()
+	if len(sigs) != len(signers) {
+		return errorsmod.Wrapf(
+			sdkerrors.ErrUnauthorized,
+			"wrong number of signers; expected %d, got %d", len(signers), len(sigs),
+		)
+	}
+
+	return nil
```

Comment:
It looks like the upper portion of this function is identical to the `ValidateBasic` defined in `types/tx/types.go`.  I think it's best to call that one first, then assert the additional conditions to avoid duplicated code.

```suggestion
func (w *wrapper) ValidateBasic() error {
	err := w.tx.ValidateBasic()
	if err != nil {
		return err
	}

	sigs := w.tx.Signatures
	signers, _ := w.GetSigners()
	if len(sigs) != len(signers) {
		return errorsmod.Wrapf(
			sdkerrors.ErrUnauthorized,
			"wrong number of signers; expected %d, got %d", len(signers), len(sigs),
		)
	}

	return nil
}
```